### DISCUSSION TITLE
How to create a query dynamically?

### DISCUSSION THREAD

**Discussion by vincent-herlemont** [diesel-rs/diesel#4381] [Q&A] [answered]
Is there a way to create a query dynamically, meaning to construct different blocks of the query following certain logic?

A simple case: I may or may not want a filter to be applied to the query, and I want to avoid writing two separate queries for this.

Here's what I naively did; however, it doesn't compile because I don't know what to put in `table.filter(??)` when there is no condition, and I have to put something to respect the typing of the `if` branches:

```rust
    let table = nodes::table;

    let query = if let Some(filter) = filter {
        table.filter(nodes::name.eq("Node2"))
    } else {
        table.filter(??)
    };

    query.load::<Node>(conn);
```

Undoubtedly my method is not correct, do you have any suggestions?

**Comment by weiznich** [comment, marked as answer]
See [the documentation of `QueryDsl::into_boxed`](https://docs.diesel.rs/2.2.x/diesel/prelude/trait.QueryDsl.html#method.into_boxed) for a solution.

**Reply by vincent-herlemont** [reply]
Thanks works perfectly!

Result
```rust
    let mut query = nodes::table.into_boxed();

    if let Some(_) = filter {
        query = query.filter(nodes::name.eq("Node2"))
    };

    query.load::<Node>(conn);
```
